lib/cpp-test/serialization/solver_gtest.cpp

```diff
@@ -0,0 +1,278 @@
+
+#include "include.h"
+
+#include "tick/solver/adagrad.h"
+#include "tick/solver/sdca.h"
+#include "tick/solver/sgd.h"
+#include "tick/solver/saga.h"
+#include "tick/solver/svrg.h"
+
+
+template <class T, class MODEL>
+void test_linear_separable(std::function<std::shared_ptr<MODEL>(
+                  const std::shared_ptr<BaseArray2d<T> > &features,
+                  const std::shared_ptr<SArray<T> > &labels)>
+                  get_model) {
+  auto features = get_features<T>();
+  auto labels = get_labels<T>();
+  ulong n_samples = features->n_rows();
+  const auto BETA = 1e-10;
+  const auto STRENGTH = (1. / n_samples) + BETA;
+  auto model = get_model(features, labels);
+  auto proxl1   = std::make_shared<TProxL1<T> >(STRENGTH, true);
+  auto proxl1w  = std::make_shared<TProxL1w<T> >(STRENGTH, get_proxl1w_weights<T>(), true);
+  auto proxl2   = std::make_shared<TProxL2<T> >(STRENGTH, 0,
+                                      model->get_n_coeffs(), 0);
+  auto proxl2sq = std::make_shared<TProxL2Sq<T> >(STRENGTH, true);
+  auto proxelas = std::make_shared<TProxElasticNet<T> >(STRENGTH, BETA / STRENGTH, 0,
+                                      model->get_n_coeffs(), 0);
+  auto proxeq   = std::make_shared<TProxEquality<T> >(STRENGTH, 0,
+                                      model->get_n_coeffs(), 0);
+  auto proxtv   = std::make_shared<TProxTV<T> >(STRENGTH, 0,
+                                      model->get_n_coeffs(), 0);
+  auto proxpos  = std::make_shared<TProxPositive<T> >(STRENGTH);
+  auto prox0    = std::make_shared<TProxZero<T> >(STRENGTH);
+  {
+    auto solver = std::make_shared<TAdaGrad<T> >(n_samples, 0, RandType::unif, 1e3, -1);
+    SCOPED_TRACE("");
+    WITH_PROX(
+      proxl1, proxl1w, proxl2, proxl2sq, proxelas, proxeq, proxtv, proxpos, prox0
+    ); 
+  }
+  {
+    auto solver = std::make_shared<TSDCA<T> >(n_samples);
+    SCOPED_TRACE("");
+    WITH_PROX(
+      proxl1, proxl1w, proxl2, proxl2sq, proxelas, proxeq, proxtv, proxpos, prox0
+    ); 
+  }
+  {
+    auto solver = std::make_shared<TSGD<T> >(n_samples, 0, RandType::unif, 1e3, -1);
+    SCOPED_TRACE("");
+    WITH_PROX(
+      proxl1, proxl1w, proxl2, proxl2sq, proxelas, proxeq, proxtv, proxpos, prox0
+    ); 
+  }
+  {
+    auto solver = std::make_shared<TSVRG<T> >(n_samples, 0, RandType::unif, 1e3, -1);
+    SCOPED_TRACE("");
+    WITH_PROX(
+      proxl1, proxl1w, proxl2, proxl2sq, proxelas, proxeq, proxtv, proxpos, prox0
+    ); 
+  }
+  {
+    auto solver = std::make_shared<TSVRG<T> >(n_samples, 0, RandType::unif, 1e3, -1);
+    SCOPED_TRACE("");
+    WITH_PROX(
+      proxl1, proxl1w, proxl2, proxl2sq, proxelas, proxeq, proxtv, proxpos, prox0
+    ); 
+  }
+}
+
+//####################### LIN REG ############################################
```

Comment:
Maybe can we put this in model_gtest.cpp?

Reply:
it's solvers tho - I removed the part from models_gtest.cpp

Reply:
ok I seem to have left it in models_gtest.cpp - this is a much more extensive set of tests though encapsulating the entire solver/model/prox group